Decouple offline hash strength from online

rnewson opened the pull request:
## Overview

Allow arbitrarily strong on-disk password hashes without impacting database request performance.

This PR introduces PBKDF2 with SHA-256 (our existing version only uses SHA-1 which is now deprecated).

This PR also introduces an in-memory cache of password hashes with a separate, and deliberately low, iteration
count. The first request for a given user is forced to perform a slow password hash check (as is any attacker that
might have access to the on-disk password hashes). Subsequent hashes are faster. Entries in the password hash
cache are time-limited, unused entries are automatically deleted, and there is a capacity bound.

## Testing recommendations

covered by automated tests.

## Related Issues or Pull Requests

N/A

## Checklist

- [x] Code is written and works correctly
- [x] Changes are covered by tests
- [x] Any new configurable parameters are documented in `rel/overlay/etc/default.ini`
- [x] Documentation changes were made in the `src/docs` folder
- [ ] Documentation changes were backported (separated PR) to affected branches

nickva reviewed (commented):
Very nice! It's about time we moved on from SHA-1.

I had a few general first impressions:

* I wonder if we could de-couple the cache mechanism from adding a sha256 pbkdf2 variant, so it would be two separate PRs?

On the cache component:
 
* Some users or services may already generate random passwords and salt values, say for API keys. Perhaps we can avoid forcing them to go through the extra ETS table lookup and an extra gen_server cast for every request.

On the built-in SHA256 PBKDF2 bit:

* Seeing as the new `crypto:pbkdf2_hmac` call in Erlang/OTP goes directly through to the C API https://github.com/erlang/otp/blob/maint-24/lib/crypto/c_src/pbkdf2_hmac.c#L60C10-L60C27, it raises red flags that this operation doesn't properly yield and hogs the main schedulers. That would be appropriate for a quick 5-10 msec call, but not for one expecting to last 100s or 1000s of msecs.

rnewson commented:
I'd considered https://github.com/esl/fast_pbkdf2/ before I noticed they'd made it part of crypto.erl. It just hadn't occurred to me that it would be worse. I will switch to fast_pbkdf2 instead, which does the NIF yieldy goop we need.

rnewson commented:
I deliberately separated into multiple commits but I feel adding pbkdf2-sha256 with high iterations without the cache would lead to problems for folks.

the password cache could be optional, I could add a toggle so admins that want to can disable it. given that the work effort for a cached item is quite small (though not as small as the single-pass sha1) I didn't think it was necessary, but it's also easy to add.

nickva commented:
> I'd considered https://github.com/esl/fast_pbkdf2/ before I noticed they'd made it part of crypto.erl. It just hadn't occurred to me that it would be worse. I will switch to fast_pbkdf2 instead, which does the NIF yieldy goop we need.

Or even upgrading our pbkdf2 could work too but yeah that one does seem to do better.

Here is an escript which shows the abnormal behavior:

```erlang
#!/usr/bin/env escript
%% -*- erlang -*-

-mode(compile).

hash() ->
    crypto:pbkdf2_hmac(sha256, <<"password">>, <<"salt">>, 1500000, 32),
    %timer:sleep(1),
    %erlang:yield(),
    hash().

spawn_hasher() ->
    spawn_link(fun() ->
        timer:sleep(round(rand:uniform() * 10)),
        hash()
    end).

main([TimeSecStr, HashersStr]) ->
    TimeSec = list_to_integer(TimeSecStr),
    Hashers = list_to_integer(HashersStr),
    HasherPids = [spawn_hasher() || _ <- lists:seq(1, Hashers)],
    io:format(" * ~p hashers started, measuring scheduler usage for ~p seconds~n", [Hashers, TimeSec]),
    T0 = erlang:monotonic_time(millisecond),
    msacc:start(TimeSec * 1000),
    T1 = erlang:monotonic_time(millisecond),
    io:format(" * killing ~p hashers~n", [Hashers]),
    [begin unlink(Pid), exit(Pid, kill) end || Pid <- HasherPids],
    io:format(" * displaying msacc stats~n", []),
    msacc:print(msacc:stats(), #{system => true}),
    io:format(" * time it took to run: ~p~n", [(T1-T0)/1000]),
    io:format(" * done~n").
```


It should be done in about 10 second but it just block and until I stopped it by hand:
```
% time ./pbkdfprocs.escript 10 4
 * 4 hashers started, measuring scheduler usage for 10 seconds
^C
./pbkdfprocs.escript 10 4  51.52s user 0.38s system 99% cpu 52.347 total
```

I'll make an OTP issue, otherwise the function, as is, is completely unusable for its intended purpose.

nickva commented:
Upstream OTP issue: https://github.com/erlang/otp/issues/7769

big-r81 commented:
Some additional infos about the iteration count: https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#pbkdf2

nickva commented:
> I deliberately separated into multiple commits but I feel adding pbkdf2-sha256 with high iterations without the cache would lead to problems for folks.
> 

The cache would be for the increased default iteration count? It would make sense to add it even for a sha1 higher iteration count. sha256 is a bit more expensive but still within the same order of magnitude as sha1.

```
> timer:tc(fun() ->  crypto:pbkdf2_hmac(sha256, <<"password">>, <<"salt">>, 10000, 32) end).
{7010,
 <<94,192,43,145,164,181,156,111,89,221,95,190,76,166,73,
   236,228,250,133,104,205,184,186,54,207,65,66,...>>}

> timer:tc(fun() ->  crypto:pbkdf2_hmac(sha, <<"password">>, <<"salt">>, 10000, 20) end).
{4951,
 <<162,194,100,97,134,130,132,116,183,84,89,26,84,124,24,
   241,50,216,141,116>>}
```

> the password cache could be optional, I could add a toggle so admins that want to can disable it. given that the work effort for a cached item is quite small (though not as small as the single-pass sha1) I didn't think it was necessary, but it's also easy to add.

The main concern there is that concurrent requests will check the single ets table, which also ends up being updated constantly with the atime `accessed` operations.  It seems if users have generated random 128 bit salts and password for  API keys, they should be able to use a sha256 or sha512 with a low iteration number to avoid an extra hash and ets table check.

rnewson commented:
@nickva I've made the cache a configurable option. for deployments that are confident they avoid basic auth they can disable it without negative consequences.

rnewson commented:
noting I just rebased after making an important internal change. the salt value for the internal cache is now the same as the one from the users doc. This ensures that the entry is not used if a user changes their password (as their salt value is re-randomised). We were already preserving the salt value on upgrade to account to avoid conflicts being introduced. If a customer updates their password themselves (by editing their user doc) a new salt is generated.

nickva commented:
> I've made the cache a configurable option. for deployments that are confident they avoid basic auth they can disable it without negative consequences

Makes sense.

Also, good idea to go with https://github.com/esl/fast_pbkdf2/ for now. I made a PR for Erlang/OTP to fix the blocking issue https://github.com/erlang/otp/pull/7770 so far it seems to be in review and not sure what's blocking it. Hoping that would make it to maint and the next 24/25/26 patch releases.

One, perhaps minor or temporary, issue with fast_pbkdf2 is that it's not FIPS-140 validated. In other words, even if the hashing code may be copy-pasted from OpenSSL library, it might not be accepted as "compliant". Not that we necessarily strive to be compliant with FIPS 140, but we do have the extra feature and mode for it, so we would then prevent users from using that feature if we go with a native hashing primitives.

rnewson commented:
hm, on FIPS-140 the fundamental algorithms (the SHA family) are invoked from openssl, which in such an environment is where the FIPS-140 "compliance" comes from. It will prevent the use of MD5 etc. not sure it's necessary to use PBKDF2 at that level, but it's a good point and needs confirmation either way.

rnewson commented:
(of course our current PBKDF2 implementation is bespoke anyway)

nickva commented:
Ah indeed, is invoking the sha* functions from OpenSSL. The macros in fast_pbkdf2.c looked a bit confusing and it was hard to tell what was happening there. And you're right, in that respect it's just as compliant or non-compliant as our previous implementation.

nickva commented on src/couch/src/couch_users_db.erl:
```diff
@@ -63,6 +64,21 @@ before_doc_update(Doc, Db, _UpdateType) ->
+    PreserveSalt = couch_util:get_value(?PRESERVE_SALT, Body, false),
+    Salt =
+        if
+            PreserveSalt ->
+                % use existing salt, if present.
+                couch_util:get_value(?SALT, Body, couch_uuids:random());
+            true ->
+                couch_uuids:random()
```
Minor nit: `if true -> ... ; true -> ... end` even if correct will look confusing.  A simple `case` might be more clear instead?

rnewson replied:
sure

nickva commented on src/couch/src/couch_password_hasher.erl:
```diff
@@ -45,29 +66,95 @@ start_link() ->
+        "iterations", 10
+    ),
+    if
+        CurrentScheme == TargetScheme andalso TargetScheme == <<"simple">> andalso
+            CurrentIterations == TargetIterations ->
+            false;
+        CurrentScheme == TargetScheme andalso TargetScheme == <<"pbkdf2">> andalso
+            CurrentPRF == TargetPRF andalso CurrentIterations == TargetIterations ->
+            false;
+        true ->
+            true
+    end.
```
minor style nit: another way to structure it could be something like

```erlang
case {TargetScheme, TargetIterations, TargetPRF} of
    {CurrentScheme, CurrentIterations, _} when CurrentScheme == <<"simple">> -> false;
    {CurrentScheme, CurrentIterations, CurrentPRF} when CurrentScheme == <<"pbkdf2">> -> false;
    {_, _, _} -> true
end
```

rnewson replied:
I'll try it but I struggled here to be clear and found the `if` clauses the clearest option.

nickva commented on src/couch/src/couch_password_hasher.erl:
```diff
@@ -21,20 +21,41 @@
+    NeedsUpgrade = needs_upgrade(UserProps),
+    InProgress = in_progress(AuthModule, UserName),
+    if
+        not IsAdmin andalso NeedsUpgrade andalso not InProgress ->
```
We upgrade user doc passwords but not admins. Would it make sense to have a way to upgrade those too?

nickva replied:
That makes sense. We could put it in a comment in the code perhaps right above `maybe_upgrade_password_hash`, just so we if we change the code it wouldn't get out of sync with a readme.md file somewhere.

rnewson replied:
I might try the admin upgrade to see what it looks like, now that I added the piece to preserve salt for upgrades I think I can mitigate any problems with the upgrade, it's just whether the fake admin doc makes the update harder, etc. but I would happily proceed with merging without it.

nickva commented on src/couch/src/couch_password_hasher.erl:
```diff
@@ -45,29 +66,95 @@ start_link() ->
+handle_info({done, AuthModule, UserName}, State) ->
+    ets:delete(?IN_PROGRESS_ETS, {AuthModule, UserName}),
+    {noreply, State};
+handle_info({'DOWN', MonRef, _, _, normal}, #state{worker_mon = MonRef} = State) ->
+    ets:delete_all_objects(?IN_PROGRESS_ETS),
+    {noreply, start_worker_loop(State)};
```
It doesn't seem like the worker would ever exit `normal`. As soon as `receive` expression returns, recurse right back to the top. We expect it to run as long as password hasher gen_server is up.

rnewson replied:
good catch, I will `_` the reason.

nickva commented on src/couch/src/couch_password_hasher.erl:
```diff
@@ -45,29 +66,95 @@ start_link() ->
+    ets:member(?IN_PROGRESS_ETS, {AuthModule, UserName}).
+
+start_worker_loop(State) ->
+    {WorkerPid, WorkerMon} = spawn_monitor(?MODULE, worker_loop, [self()]),
```
`spawn_monitor` will keep the worker up even after its parent crashes. Should we `spawn_link` it instead and then in a `terminate/2` function to `unlink` and `kill` it?

If we expect it to crash often we could set the parent to trap exits and then react to an `{'EXIT', ...}` to log an error, cleanup the table and restart it again.

rnewson replied:
I think spawn_link and just letting worker_loop's death kill couch_password_hasher seems best / least surprising. will change.

rnewson replied:
updated

nickva commented on src/couch/src/couch_passwords.erl:
```diff
@@ -69,106 +68,74 @@ get_unhashed_admins() ->
     is_integer(Iterations),
-    Iterations > 0
+    Iterations > 0,
+    KeyLen > 0
```
If we check a value like `KeyLen` for `> 0` it's usually a good practice to always assert it's an integer as well.

rnewson replied:
done

nickva commented on src/couch/src/couch_passwords.erl:
```diff
@@ -69,106 +68,74 @@ get_unhashed_admins() ->
     is_integer(Iterations),
-    Iterations > 0
+    Iterations > 0,
+    KeyLen > 0
```
Let's assert KeyLen is an integer, too

rnewson replied:
done

nickva commented on src/couch/src/couch_passwords.erl:
```diff
@@ -69,106 +68,74 @@ get_unhashed_admins() ->
     is_integer(Iterations),
-    Iterations > 0
+    Iterations > 0,
+    KeyLen > 0
```
Let's assert KeyLen is an integer.

rnewson replied:
done

nickva commented on src/couch/src/couch_passwords_cache.erl:
```diff
@@ -0,0 +1,65 @@
+    AuthModule :: atom(), UserName :: binary(), Password :: binary(), Salt :: binary()
+) ->
+    not_found | boolean().
+authenticate(AuthModule, UserName, Password, Salt) ->
```
Instead of just specs, which are great, it might make sense to also assert them as guards?

rnewson replied:
done

nickva commented on src/couch/src/couch_passwords_cache.erl:
```diff
@@ -0,0 +1,65 @@
+    ets_lru:insert(?CACHE, {AuthModule, UserName, Salt}, hash(Password, Salt)).
+
+hash(Password, Salt) ->
+    fast_pbkdf2:pbkdf2(sha256, Password, Salt, _Iterations = 5, _KeyLen = 32).
```
Not sure if the prf, iterations and keylen would look better as macros, to stand out at the top as parameters of this module. If you prefer them as is, that's fine too, it's more concise perhaps.

rnewson replied:
This was my way of showing that the in-memory hashing is not configurable by design.

nickva replied:
Well, it wouldn't be configurable anyway, but the idea is to make them more visible as they are parameters for this cache module. Perhaps with a comment about why iterations is 5 and maybe what threat model we're assuming. If I were auditing the code ` _Iterations = 5` would jump out immediately as a red flag.

Since  fast_pbkdf2 is an external library let's isolate all the calls to it in `couch_passwords` or other hash utility module if we have one. If my otp pbdkf2 unlocking pr merges, we'd have less places to replace.

nickva commented on src/couch/src/couch_passwords_cache.erl:
```diff
@@ -0,0 +1,65 @@
+% public api
+-export([authenticate/4, insert/4]).
+
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
```
Do we use this call? This module doesn't look like a gen_server?

rnewson replied:
removed

nickva commented on src/couch/src/couch_passwords_cache.erl:
```diff
@@ -0,0 +1,65 @@
+
+-module(couch_passwords_cache).
+
+-define(CACHE, couch_passwords_cache_lru).
```
Would it simplify naming a bit if we just used ?MODULE in this module and `couch_passwords_cache` in `couch_primary_sup`? We already know it's an LRU based on the ets_lru insert/lookup_d calls below.

rnewson replied:
good catch, this existed before I switched to ets_lru

rnewson commented:
sorry, I've modified this yet again. Notably I increase the default iterations to 50,000. I override this in a test to avoid adding a timer:sleep hack.

I changed needs_upgrade to _not_ have defaults for password_scheme or iterations. Instead I use couch_util:get_value/2 and so we get `undefined` if these items are missing. If either are missing the function will return true and we'll then do the upgrade. Any document without these fields is broken and needs fixing (a user/admin that removed them by accident is the only path). I much prefer this to yet another place with a `sha` or `10` default. It is quite _wrong_ to put defaults there anyway. The default for scheme, prf, iterations only apply when _creating_ new password hashes. Existing hashes must have their actual parameters supplied (as fields for user docs and packed into the config value for admins).

nickva commented on rebar.config.script:
```diff
@@ -157,7 +157,8 @@ DepDescs = [
 {meck,             "meck",             {tag, "0.9.2"}},
-{recon,            "recon",            {tag, "2.5.3"}}
+{recon,            "recon",            {tag, "2.5.3"}},
+{fast_pbkdf2,      {url, "https://github.com/esl/fast_pbkdf2"}, {tag, "1.0.2"}}
```
Since we're planning a CouchDB release soon-ish, wonder if it makes sense to just vendor it for now?

At some point https://github.com/erlang/otp/pull/7770 will merge then if we made a release with https://github.com/esl/fast_pbkdf2's mirror we'd have to archive it and keep it essentially forever.

rnewson replied:
by "just vendor it" you mean "keep pointing at the /esl/ repo we don't control"? I don't think it's an option per ASF rules.

nickva replied:
No I meant include it in the applications directory alongside ets_lru, exxhash and others.

rnewson replied:
hmmmm interesting

nickva commented on src/fast_pbkdf2/.gitignore:
```diff
@@ -0,0 +1,30 @@
+.rebar3
```
Noticed from rebar compiler, it leaves the `src/fast_pbkdf2/compile_commands.json` behind as well. Can probably add it to .gitignore. Either this one or the top level one.

```
% git status

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	src/fast_pbkdf2/compile_commands.json
```

nickva replied:
Noticed from rebar compiler, it leaves the `src/fast_pbkdf2/compile_commands.json` behind as well. Can probably add it to .gitignore. Either this one or the top level one.

```
% git status

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	src/fast_pbkdf2/compile_commands.json
```

nickva commented on src/fast_pbkdf2/rebar.config:
```diff
@@ -0,0 +1,58 @@
+        % From files
+        ["c_src/*.c"],
+        % Using options
+        [ {env, [{"CFLAGS", "$CFLAGS -std=c99 -O3 -g -Wall -Wextra -fPIC"},
```
A full CI Jenkins showed that all the architecture and OSes looked good except macos. I think it may just need the openssl dev library installed:

https://ci-couchdb.apache.org/blue/organizations/jenkins/jenkins-cm1%2FFullPlatformMatrix/detail/jenkins-decouple_offline_hash_strength_from_online/1/pipeline/82/

```
[2023-11-07T22:26:54.619Z] Compiling /Users/jenkins2/workspace/ffline_hash_strength_from_online/macos/build/src/fast_pbkdf2/c_src/fast_pbkdf2.c
[2023-11-07T22:26:54.619Z] /Users/jenkins2/workspace/ffline_hash_strength_from_online/macos/build/src/fast_pbkdf2/c_src/fast_pbkdf2.c:35:10: fatal error: 'openssl/sha.h' file not found
[2023-11-07T22:26:54.619Z] #include <openssl/sha.h>
[2023-11-07T22:26:54.619Z]          ^~~~~~~~~~~~~~~
[2023-11-07T22:26:54.619Z] 1 error generated.
[2023-11-07T22:26:54.619Z] ERROR: compile failed while processing /Users/jenkins2/workspace/ffline_hash_strength_from_online/macos/build/src/fast_pbkdf2: rebar_abort
[2023-11-07T22:26:54.619Z] make: *** [couch] Error 1
```

nickva commented on dev/run:
```diff
@@ -571,8 +572,8 @@ def hashify(pwd, salt=COMMON_SALT, iterations=10, keylen=20):
-    derived_key = pbkdf2_hex(pwd, salt, iterations, keylen)
-    return "-pbkdf2-%s,%s,%s" % (derived_key, salt, iterations)
+    derived_key = pbkdf2_hex(pwd, salt, iterations, keylen, hashfunc=hashlib.sha256)
+    return "-pbkdf2:sha256-%s,%s,%s" % (derived_key, salt, iterations)
```
I couldn't start a local dev n=1 node cluster for some reason.

```
 % ./dev/run --admin=adm:pass -n 1
[ * ] Setup environment ... ok
[ * ] Ensure CouchDB is built ... ok
[ * ] Ensure Erlang boot script exists ... ok
[ * ] Prepare configuration files ... ok
[ * ] Start node node1 ... ok
[ * ] Check node at http://127.0.0.1:15984/ ... ok
[ * ] Running cluster setup ... failed: b'{"error":"unauthorized","reason":"Name or password is incorrect."}\n'
Traceback (most recent call last):
  File "/Users/nvatama/asf/./dev/run", line 934, in <module>
    main()
  File "/Users/nvatama/asf/./dev/run", line 93, in main
    startup(ctx)
  File "/Users/nvatama/asf/./dev/run", line 593, in startup
    cluster_setup(ctx)
  File "/Users/nvatama/asf/./dev/run", line 73, in wrapper
    res = func(*args, **kwargs)
  File "/Users/nvatama/asf/./dev/run", line 743, in cluster_setup
    finish_cluster(lead_port, *ctx["admin"])
  File "/Users/nvatama/asf/./dev/run", line 829, in finish_cluster
    assert resp.status in (201, 400), resp.read()
AssertionError: b'{"error":"unauthorized","reason":"Name or password is incorrect."}\n'
```

The logs showed:

```
[notice] 2023-11-07T23:19:16.524311Z node1@127.0.0.1 <0.494.0> -------- Missing system database _users
[notice] 2023-11-07T23:19:17.254909Z node1@127.0.0.1 <0.430.0> e86161454e 127.0.0.1:15984 127.0.0.1 undefined GET / 200 ok 63
[notice] 2023-11-07T23:19:17.276089Z node1@127.0.0.1 <0.431.0> acf614ed19 Setup: [{<<"action">>,<<"enable_cluster">>},{<<"bind_address">>,<<"0.0.0.0">>},{<<"username">>,<<"adm">>},{<<"password">>,<<"****">>},{<<"node_count">>,1}]

[notice] 2023-11-07T23:19:17.278581Z node1@127.0.0.1 <0.141.0> -------- config: [admins] adm set to '****' for reason nil
[notice] 2023-11-07T23:19:17.281576Z node1@127.0.0.1 <0.141.0> -------- config: [chttpd] bind_address set to 0.0.0.0 for reason nil
[notice] 2023-11-07T23:19:17.282793Z node1@127.0.0.1 <0.141.0> -------- config: [cluster] n set to 1 for reason nil
[error] 2023-11-07T23:19:17.282809Z node1@127.0.0.1 <0.427.0> -------- Supervisor chttpd_sup had child chttpd started with chttpd:start_link() at <0.429.0> exit with reason normal in context child_terminated
[error] 2023-11-07T23:19:17.282840Z node1@127.0.0.1 <0.427.0> -------- Supervisor chttpd_sup had child chttpd started with chttpd:start_link() at <0.429.0> exit with reason normal in context child_terminated
[notice] 2023-11-07T23:19:17.284662Z node1@127.0.0.1 <0.431.0> acf614ed19 127.0.0.1:15984 127.0.0.1 adm POST /_cluster_setup 201 ok 26
[notice] 2023-11-07T23:19:17.286358Z node1@127.0.0.1 <0.585.0> 53d7202075 127.0.0.1:15984 127.0.0.1 undefined POST /_cluster_setup 401 ok 0
[info] 2023-11-07T23:19:17.288890Z node1@127.0.0.1 <0.66.0> -------- SIGTERM received - shutting down
[info] 2023-11-07T23:19:17.288923Z node1@127.0.0.1 <0.66.0> -------- SIGTERM received - shutting down
```
I couldn't yet tell why though. I'll try adding more debug statements in there to figure it out

nickva replied:
`-n 3` seems to work though, so it's a bit strange.

rnewson replied:
one note here: an admin cred can't be used until it's hashed, so if iterations is high (or the machine is slow), dev/run doesn't boot as it doesn't wait long enough for the hashing. This is a latent issue, it's just exposed more with the new defaults (50_000 in default.ini, which is why the Makefile overrides it to 9, so those dev/run's boot up quickly).

nickva replied:
That explains it, thanks! As part of that setup we re-set the admin credentials by calling setup.erl [set_admin](https://github.com/apache/couchdb/blob/main/src/setup/src/setup.erl#L172). That sets it to plaintext password and then immediately returns. The config:set then asynchronously kicks of the couch_server admin re-hasher.

nickva reviewed (changes requested):
I noticed a few build and dev/cluster related issues. Added them to the review. 

I ran full CI: https://ci-couchdb.apache.org/blue/organizations/jenkins/jenkins-cm1%2FFullPlatformMatrix/detail/jenkins-decouple_offline_hash_strength_from_online/1/pipeline/82/ worked well for most architecture but macos and @big-r81 indicates that Windows may need an updated. For macos might just be a simple brew openssl dev install.

big-r81 commented:
First compile run:

```
==> fast_pbkdf2 (compile)
Compiling c:/Users/couchdb/Documents/couchdb/src/fast_pbkdf2/c_src/fast_pbkdf2.c
ERROR: compile failed while processing c:/Users/couchdb/Documents/couchdb/src/fast_pbkdf2: rebar_abort
Microsoft (R) C/C++ Optimizing Compiler Version 19.37.32825 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

cl : Command line error D8021 : invalid numeric argument '/Wextra'
make: *** [Makefile.win:118: couch] Error 1
```
So I patched `rebar.config` of `fast_pbkdf2` to:

```diff
diff --git a/src/fast_pbkdf2/rebar.config b/src/fast_pbkdf2/rebar.config
index a0ef529ca..871f8a74f 100644
--- a/src/fast_pbkdf2/rebar.config
+++ b/src/fast_pbkdf2/rebar.config
@@ -39,7 +39,8 @@
         % From files
         ["c_src/*.c"],
         % Using options
-        [ {env, [{"CFLAGS", "$CFLAGS -std=c99 -O3 -g -Wall -Wextra -fPIC"},
+        [ {env, [{"(linux|solaris|darwin|freebsd)", "CFLAGS", "$CFLAGS -std=c99 -O3 -g -Wall -Wextra -fPIC"},       
+                 {"win32", "CFLAGS", "$CFLAGS /O2 /DNDEBUG /Wall"},
                  {"LDLIBS", "$LDLIBS -lcrypto"},
                  {"DRV_LINK_TEMPLATE", "$DRV_LINK_TEMPLATE $LDLIBS"}]}]
      }
```

Second compile run:

```
==> fast_pbkdf2 (compile)
Compiling c:/Users/couchdb/Documents/couchdb/src/fast_pbkdf2/c_src/fast_pbkdf2.c
Microsoft (R) C/C++ Optimizing Compiler Version 19.37.32825 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

fast_pbkdf2.c
c:\tools\erl-24.3.4.10\erts-12.3.2.10\include\erl_drv_nif.h(40): warning C4820: '<unnamed-tag>': '4' bytes padding added after data member 'dirty_scheduler_support'
c:\tools\erl-24.3.4.10\erts-12.3.2.10\include\erl_drv_nif.h(156): warning C4820: '_SysIOVec': '4' bytes padding added after data member 'iov_len'
c:\tools\erl-24.3.4.10\erts-12.3.2.10\include\erl_nif.h(123): warning C4820: 'enif_func_t': '4' bytes padding added after data member 'arity'
c:\tools\erl-24.3.4.10\erts-12.3.2.10\include\erl_nif.h(125): warning C4820: 'enif_func_t': '4' bytes padding added after data member 'flags'
c:\tools\erl-24.3.4.10\erts-12.3.2.10\include\erl_nif.h(134): warning C4820: 'enif_entry_t': '4' bytes padding added after data member 'num_of_funcs'
c:\tools\erl-24.3.4.10\erts-12.3.2.10\include\erl_nif.h(147): warning C4820: 'enif_entry_t': '4' bytes padding added after data member 'options'
c:\tools\erl-24.3.4.10\erts-12.3.2.10\include\erl_nif.h(215): warning C4820: '<unnamed-tag>': '4' bytes padding added after data member 'members'
c:\tools\erl-24.3.4.10\erts-12.3.2.10\include\erl_nif.h(279): warning C4820: 'erl_nif_io_vec': '4' bytes padding added after data member 'iovcnt'
c:\tools\erl-24.3.4.10\erts-12.3.2.10\include\erl_nif.h(287): warning C4820: 'erl_nif_io_vec': '4' bytes padding added after data member 'flags'
c:/Users/couchdb/Documents/couchdb/src/fast_pbkdf2/c_src/fast_pbkdf2.c(35): fatal error C1083: Cannot open include file: 'openssl/sha.h': No such file or directory
ERROR: compile failed while processing c:/Users/couchdb/Documents/couchdb/src/fast_pbkdf2: rebar_abort
make: *** [Makefile.win:118: couch] Error 1
```

Windows has problems with OpenSSL too. OpenSSL is installed via `vcpkg` and should be automatically available (header files etc...). Need some more investigations.

big-r81 commented:
Solved the "OpenSSL not found" error. Installed the wrong architecture (x86 instead x64).
Needed to tweak the 'rebar.config' again, that the liibrary linking is working:

```diff
diff --git a/src/fast_pbkdf2/rebar.config b/src/fast_pbkdf2/rebar.config
index a0ef529ca..486434801 100644
--- a/src/fast_pbkdf2/rebar.config
+++ b/src/fast_pbkdf2/rebar.config
@@ -39,8 +39,10 @@
         % From files
         ["c_src/*.c"],
         % Using options
-        [ {env, [{"CFLAGS", "$CFLAGS -std=c99 -O3 -g -Wall -Wextra -fPIC"},
-                 {"LDLIBS", "$LDLIBS -lcrypto"},
+        [ {env, [{"(linux|solaris|darwin|freebsd)", "CFLAGS", "$CFLAGS -std=c99 -O3 -g -Wall -Wextra -fPIC"},
+                 {"win32", "CFLAGS", "$CFLAGS /O2 /DNDEBUG /Wall"},
+                 {"(linux|solaris|darwin|freebsd)", "LDLIBS", "$LDLIBS -lcrypto"},
+                 {"win32", "LDLIBS", "$LDLIBS libcrypto.lib"},
                  {"DRV_LINK_TEMPLATE", "$DRV_LINK_TEMPLATE $LDLIBS"}]}]
      }
  ]}.
```

I think I will create a PR in the `fast_pbkdf2` repo...

rnewson commented:
I'm open to not importing fast_pbkdf2 and just enhancing our home-grown pbkdf2 implementation to allow sha-256, it's trivial to do. Since we're planning to use the native one if it gets fixed anyway, perhaps that's the best move.

nickva commented on src/couch/src/couch_httpd_auth.erl:
```diff
@@ -662,10 +643,14 @@ authenticate(Pass, UserProps) ->
             <<"pbkdf2">> ->
+                PRF = couch_util:get_value(<<"pbkdf2_prf">>, UserProps, <<"sha">>),
+                verify_prf(PRF),
                 Iterations = couch_util:get_value(<<"iterations">>, UserProps, 10000),
```
Since we bumped iterations to 50000, should this be updated as well?

rnewson replied:
yes, not sure how I missed that. thanks

rnewson replied:
hm, seconds thoughts on this. the 10000 setting is quite old and for the case where UserProps has pbkdf2 hash but has somehow lost its "iterations" property. Frankly having a default here makes no sense, so perhaps it should be a get_value/2 and then verify_iterations tweaked to fail for `undefined`?

The get_value/3 for pbkdf2_prf on the other hand _is_ right. existing pbkdf2 hashes will be hashed with `sha` and won't have this new field.

nickva replied:
That is an odd one... did we ever make a release which always assumed a hard-code iterations value of 10000, if there is a slight chance we did, we should probably keep the value as is. If you recall better and there is no chance we'd actually rely on that default 10000 then maybe doing the `undefined` or `nil` verification would make sense.

nickva commented on src/couch/src/couch_primary_sup.erl:
```diff
@@ -23,7 +23,18 @@ init([]) ->
+                {ets_lru, start_link, [
+                    couch_passwords_cache_lru,
+                    [
+                        {max_objects,
+                            config:get_integer("couch_passwords_cache", "max_objects", 10_000)},
+                        {max_lifetime,
+                            config:get_integer("couch_passwords_cache", "max_lifetime", 360_000)}
```
Let's add these to do the `default.ini`. We already have a section there for `enable = true|false`.

nickva replied:
For max_lifetime was the intention to make it 6 minutes or an hour? The unit it in milliseconds so if it should be an hour it would be `3_600_000`.

nickva commented on src/couch/src/couch_passwords_cache.erl:
```diff
@@ -0,0 +1,69 @@
+insert(AuthModule, UserName, Password, Salt) when
+    is_atom(AuthModule), is_binary(UserName), is_binary(Password), is_binary(Salt)
+->
+    ets_lru:insert(?CACHE, {AuthModule, UserName, Salt}, hash(Password, Salt)).
```
If couch_passwords_cache is not enable, we shouldn't insert into the cache?

nickva reviewed (changes requested):
I had created a script to make a bunch of users and then use them to do _bulk_docs updates concurrently. 

The default settings worked really nice. The user docs would be updated and cached properly.

https://gist.github.com/nickva/e733156c6fefdc94f6bd9051a456d9e5

Ran as:
```
./multiuser.py --users-scheme=simple --users-salt="abc" --users-hash-sha1 --tries 100
```

However then I configured caching parameters, disabled caching and auto-updates couch app crashed with a badmatch in ets_lru.

I used these parameters:

```ini
[chttpd_auth]
upgrade_hash_on_auth = false

[couch_passwords_cache]
max_lifetime = 5000
max_objects = 5
enable = false
```

I noticed that the cache was still populated with entries despite it being disabled and there was this crash in the logs:

```
[error] 2023-11-09T04:31:23.807375Z node1@127.0.0.1 <0.3506.1> -------- gen_server couch_passwords_cache_lru terminated with reason: no match of right hand value [] at ets_lru:drop_lru/2(line:289) <= ets_lru:trim/1(line:236) <= ets_lru:handle_info/2(line:213) <= gen_server:try_dispatch/4(line:695) <= gen_server:handle_msg/6(line:771) <= proc_lib:init_p_do_apply/3(line:226)
  last msg: redacted
     state: {st,couch_passwords_cache_lru_objects,couch_passwords_cache_lru_atimes,couch_passwords_cache_lru_ctimes,5,undefined,5000,millisecond}
    extra: []
[error] 2023-11-09T04:31:23.807461Z node1@127.0.0.1 <0.3506.1> -------- gen_server couch_passwords_cache_lru terminated with reason: no match of right hand value [] at ets_lru:drop_lru/2(line:289) <= ets_lru:trim/1(line:236) <= ets_lru:handle_info/2(line:213) <= gen_server:try_dispatch/4(line:695) <= gen_server:handle_msg/6(line:771) <= proc_lib:init_p_do_apply/3(line:226)
  last msg: redacted
     state: {st,couch_passwords_cache_lru_objects,couch_passwords_cache_lru_atimes,couch_passwords_cache_lru_ctimes,5,undefined,5000,millisecond}
    extra: []
```

It's possible that the cache settings were too aggressive, I'll try tomorrow with larger values. I was mainly testing the case when we hit the cache limits and/or disable the cache feature.

Update: this turned out to be a genuine bug in ets_lru. It is now fixed in https://github.com/apache/couchdb/pull/4837

nickva reviewed (commented):
This PR is in solid shape. There are few minor nits about config values and what to do about that default 10000 iteration case.

I tested it with my script with / without the cache and with the upgrade / without the upgrade option on a single node. Everything worked great. The cache works as intended. Well done!

When trying to run it on a 3 node cluster, I did notice a good amount of conflict exceptions. Not sure what we can do there without making it too complicated. Try to do it on one node only (the first node of a shard range), or something like we do for replication jobs and have an "owner" for each job based on where the doc shard lives?

```
[notice] 2023-11-09T22:49:13.624493Z node1@127.0.0.1 <0.268.0> -------- upgrading stored password hash for 'usr878'
[error] 2023-11-09T22:49:13.625140Z node1@127.0.0.1 emulator -------- Error in process <0.30357.2> on node 'node1@127.0.0.1' with exit value:
{{nocatch,conflict},[{fabric,update_doc,3,[{file,"src/fabric.erl"},{line,351}]},{chttpd_auth_cache,'-update_user_creds/3-fun-0-',1,[{file,"src/chttpd_auth_cache.erl"},{line,64}]}]}

[error] 2023-11-09T22:49:13.625280Z node1@127.0.0.1 emulator -------- Error in process <0.30357.2> on node 'node1@127.0.0.1' with exit value:
{{nocatch,conflict},[{fabric,update_doc,3,[{file,"src/fabric.erl"},{line,351}]},{chttpd_auth_cache,'-update_user_creds/3-fun-0-',1,[{file,"src/chttpd_auth_cache.erl"},{line,64}]}]}
```

rnewson commented:
I'd like to suppress the conflict errors since those are benign. Each node might elect to hash and update the user doc when it gets a request. however, by design of this PR, they will attempt to make the same update. the same initial _rev, and the same new fields (iterations, derived_key). so the conflicts should be clean, in that the update is rejected at all copies and does not introduce a new branch in the doc.

rnewson commented:
I can't trigger the conflict cases locally but I've added a `catch` on the front `couch_password_hasher`'s call to `AuthModule:update_user_creds`.

nickva reviewed (approved):
+1
